examples/with-next/app/layout.tsx

```diff
@@ -0,0 +1,29 @@
+import type { Metadata } from 'next'
+import { Inter } from 'next/font/google'
+import { MSWProvider } from './msw-provider'
+
+if (process.env.NEXT_RUNTIME === 'nodejs') {
+  const { server } = require('../mocks/node')
+  server.listen()
+}
+
+const inter = Inter({ subsets: ['latin'] })
+
+export const metadata: Metadata = {
+  title: 'Create Next App',
+  description: 'Generated by create next app',
+}
+
+export default function RootLayout({
+  children,
+}: Readonly<{
+  children: React.ReactNode
+}>) {
+  return (
+    <html lang="en">
+      <body className={inter.className}>
+        <MSWProvider>{children}</MSWProvider>
+      </body>
+    </html>
+  )
```

Comment:
Won't it be an issue that we need to change our application behaviour to utilise client components only to take advantage of mswjs

Reply:
So `MSWProvider` being a client component forces the rest of its children tree to be client components only? Is that the issue?

Reply:
@kettanaito It won't force the components within `MSWProvider` to be client components unless you import them directly into a file with the `"use client"` directive—[server components can be passed as `children` or other props to client components](https://nextjs.org/docs/app/building-your-application/rendering/composition-patterns#supported-pattern-passing-server-components-to-client-components-as-props).

However, I think passing `null` to `fallback` in the `Suspense` boundary in `msw-provider.tsx` results in a blank page until the promise resolves on the client-side. What we've done in our projects is pass `children` as the `fallback` so the original tree is rendered while the promise resolves. I have no idea if this has performance implications at scale, but it's worked for our purposes.

Reply:
@dbk91 You're absolutely right, sorry I missed it. My point is incorrect in that case

Reply:
@dbk91, 

> However, I think passing null to fallback in the Suspense boundary in msw-provider.tsx results in a blank page until the promise resolves on the client-side.

That is intended. Your app _mustn't_ render until MSW is activated. If it does, it may fire HTTP requests and nothing will be there to intercept them. You must use `null` as the suspense fallback.